src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/PoolOfThings.cs

```diff
@@ -0,0 +1,374 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WCFClientStressTests
+{
+    // Pooling of WCF channel factories and channels is a commonly reccomended practice due to its perf benefits.
+    // However with pooling come more race condition-like scenarios such as using an already faulted channel or factory.
+    // 
+    // The number of such scenarios is rather significant:
+    // - create and use a pool of channels using a pool of channel factories
+    // - replace and close channels under stress
+    // - replace and close channel factories under stress
+    // - manipulate a pooled channel or a pooled factory state while others use them
+    // 
+    // PoolOfThings and PoolOfAsyncThings classes help to organize generic pools of objects
+    // FactoryAndPoolOfItsObjects and FactoryAndPoolOfItsAsyncObjects classes help to organize pools of pools of objects
+    // 
+
+    public class PoolOfThings<T> : IDisposable where T : class
+    {
+        private T[] _thePool;
+        private int _maxSize;
+
+        private Func<T> _createInstance;
+        private Action<T> _destroyInstance;
+
+        /// <summary>
+        /// ctor taking sync versions of delegates
+        /// </summary>
+        /// <param name="maxSize"> number of pooled instances </param>
+        /// <param name="createInstance"> delegate to create a new instance; can return null in which case the pool will store null until DestoryAllPooledInstances is called </param>
+        /// <param name="destroyInstance"></param>
+        public PoolOfThings(int maxSize, Func<T> createInstance, Action<T> destroyInstance)
+        {
+            _thePool = new T[maxSize];
+            _maxSize = maxSize;
+            _createInstance = createInstance;
+            _destroyInstance = destroyInstance;
+        }
+
+        public void Dispose()
+        {
+            DestoryAllPooledInstances();
+        }
+
+        // Replaces the pool of channels with a new one and destory all instances in the previous pool
+        public void DestoryAllPooledInstances()
+        {
+            // Get the new pool ready
+            T[] newPool = new T[_maxSize];
+
+            // Replace the pool
+            var oldPool = _thePool;
+            if (Interlocked.CompareExchange(ref _thePool, newPool, oldPool) != oldPool)
+            {
+                // somebody beat us - they will be responsible for closing the old _channelsPool they replaced
+            }
+            else
+            {
+                // we are the one who replaced the old _thePool - we're responsible for cleaning it up
+                DestroyAllPooledInstanciesImpl(oldPool, _destroyInstance);
+            }
+        }
+
+        public IEnumerable<T> GetAllPooledInstances()
+        {
```

Comment:
Not for this PR but future consideration.  Returning an IEnumerable<T> means the values might be different every time someone evaluates it (e.g 2 Linq lambdas in a row might have different results).  If that is desired behavior, okay, but it might be better to do a .ToArray() at some point to give callers an IEnumerable that does not change